### ISSUE TITLE
Can Scatter Be Reproducible?

### ISSUE THREAD

**Issue by yuxiang-guo** [rusty1s/pytorch_scatter#226]
I used scatter function to implement GCN like this: 

re = scatter(embed, index, dim=0, out=None, dim_size=embed.size, reduce='mean')

I found that even though the input (embed and index) are the same when I run the code twice, the output are still different. So I want to know if the scatter method has any random seed, which lead to the result cannot be reproduced? How can I make the result of scatter deterministic? Thank you very much.

**Comment by rusty1s** [comment]
Scatter is a non-deterministic operation by design since it makes use of atomic operations in which the order of aggregation is non-deterministic, leading to minor numerical differences. As an alternative, you can make use of the `segment_csr` operation of `torch_scatter`.

**Comment by yuxiang-guo** [comment]
> Scatter is a non-deterministic operation by design since it makes use of atomic operations in which the order of aggregation is non-deterministic, leading to minor numerical differences. As an alternative, you can make use of the `segment_csr` operation of `torch_scatter`.


Can `segment_csr` operation replace the `scatter' method? How can I use `segment_csr`  to get the same result with `scatter' operation?

**Comment by yuxiang-guo** [comment]
> Scatter is a non-deterministic operation by design since it makes use of atomic operations in which the order of aggregation is non-deterministic, leading to minor numerical differences. As an alternative, you can make use of the `segment_csr` operation of `torch_scatter`.

Thanks very much. But how can I get the same result by using the `segment_csr` operation instead of `scatter' operation?

**Comment by rusty1s** [comment]
`segment_csr` expects indices to be sorted, and the `ptr` denotes the compressed index representation similar to a sparse matrix CSR representation:
```python
index = torch.tensor([0, 0, 0, 1, 2, 2])
ptr = torch.tensor([0, 3, 4, 6])

scatter(x, index, dim=0) == segment_csr(x, ptr)
```

**Comment by yuxiang-guo** [comment]
> `segment_csr` expects indices to be sorted, and the `ptr` denotes the compressed index representation similar to a sparse matrix CSR representation:
> 
> ```python
> index = torch.tensor([0, 0, 0, 1, 2, 2])
> ptr = torch.tensor([0, 3, 4, 6])
> 
> scatter(x, index, dim=0) == segment_csr(x, ptr)
> ```

Thank you very much!

**Comment by yuxiang-guo** [comment]
> `segment_csr` expects indices to be sorted, and the `ptr` tensor denotes the compressed index representation similar to a sparse matrix CSR representation:
> 
> ```python
> index = torch.tensor([0, 0, 0, 1, 2, 2])
> ptr = torch.tensor([0, 3, 4, 6])
> 
> scatter(x, index, dim=0) == segment_csr(x, ptr)
> ```

By the way, as for this case:

```python
index = torch.tensor([0,0,2])
scatter(x, index, dim_size=3, dim=0)
```
how can it be replaced with `segment_csr`

**Comment by yuxiang-guo** [comment]
> `segment_csr` expects indices to be sorted, and the `ptr` tensor denotes the compressed index representation similar to a sparse matrix CSR representation:
> 
> ```python
> index = torch.tensor([0, 0, 0, 1, 2, 2])
> ptr = torch.tensor([0, 3, 4, 6])
> 
> scatter(x, index, dim=0) == segment_csr(x, ptr)
> ```

By the way, if I cannot guarantee the indices are in order. And sorting is too time-consuming. How can I do or is there any other method that can achieve the same result as `scatter'?

**Comment by rusty1s** [comment]
If you cannot guarantee that indices are in order, there exists no parallelized operation that can group elements while still being deterministic. This is just the natural imprecision of floating point arithmetic.
